<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="tbor">

  <xacro:property name="base_width" value="0.375"/>
  <xacro:property name="base_height" value="0.1375"/>
  <xacro:property name="pi" value="3.14159"/>
  <xacro:property name="base_len" value="0.4"/>
  <xacro:property name="wheel_radius" value="0.075"/>
  <xacro:property name="base_wheel_gap" value="0.007"/>
  <xacro:property name="wheel_separation" value="0.4"/>
  <xacro:property name="wheel_joint_offset_x" value="0.21"/>
  <xacro:property name="wheel_joint_offset_y" value="0.140673"/>
  <xacro:property name="caster_wheel_radius" value="0.00255"/>


  <xacro:property name="caster_wheel_radius" value="0.0225"/>
  <xacro:property name="caster_wheel_mass" value="0.001"/>
  <xacro:property name="caster_wheel_joint_offset" value="0.038"/>

  <xacro:macro name="cylinder_inertia" params="m r h">
    <inertial>
      <mass value="${m}"/>
      <inertia ixx="${m*(3*r*r+h*h)/12}" ixy = "0" ixz = "0" iyy="${m*(3*r*r+h*h)/12}" iyz = "0" izz="${m*r*r/2}"/> 
    </inertial>
  </xacro:macro>

  <xacro:macro name="box_inertia" params="m w h d">
    <inertial>
      <mass value="${m}"/>
      <inertia ixx="${m / 12.0 * (d*d + h*h)}" ixy="0.0" ixz="0.0" iyy="${m / 12.0 * (w*w + h*h)}" iyz="0.0" izz="${m / 12.0 * (w*w + d*d)}"/>
    </inertial>
  </xacro:macro>

  <xacro:macro name="sphere_inertia" params="m r">
    <inertial>
      <mass value="${m}"/>
      <inertia ixx="${2.0*m*(r*r)/5.0}" ixy="0.0" ixz="0.0" iyy="${2*m*(r*r)/5}" iyz="0.0" izz="${2.0*m*(r*r)/5.0}"/>
    </inertial>
  </xacro:macro>



  <!--Base Footprint-->
  <link name="base_footprint1">
    <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
            <box size="0.001 0.001 0.001" />
        </geometry>
    </visual>
  </link>


  <!--Base link-->
  <link name="chassis">
    <xacro:box_inertia m="5" w="${base_len}" h="${base_width}" d="${base_height}"/>
    <visual>
      <origin xyz="0 0 0" rpy="0 ${pi} 0"/>
      <geometry>
        <mesh filename="/home/jainam/catkin_ws/src/test3/meshes/chassis_x5.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 ${pi} 0"/>
      <geometry>
        <mesh filename="/home/jainam/catkin_ws/src/test3/meshes/chassis_x5.dae"/>
      </geometry>
    </collision>
  </link>

  <!--base_link to base_footprint Joint-->
  <joint name="base_link_joint" type="fixed">
    <origin xyz="0 0 ${wheel_radius + 0.038441}" rpy="0 0 0" />
    <parent link="base_footprint1"/>
    <child link="chassis" />
  </joint>


    <link name="left_wheel">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="/home/jainam/catkin_ws/src/test3/meshes/wheel_x5.dae"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="/home/jainam/catkin_ws/src/test3/meshes/wheel_x5.dae"/>
        </geometry>
      </collision>
      <xacro:cylinder_inertia m="5" r="${wheel_radius}" h="0.085"/>
    </link>

    <joint name="left_wheel_joint1" type="continuous">
      <axis xyz="0 1 0" rpy="0 0 0" />
      <parent link="chassis"/>
      <child link="left_wheel"/>
      <origin xyz="${wheel_joint_offset_x} -${wheel_joint_offset_y} -0.038441" rpy="0 0 ${pi/2}"/>
    </joint>

        <link name="right_wheel">
      <visual>
        <origin xyz="0 0 0" rpy="${pi} 0 0"/>
        <geometry>
          <mesh filename="/home/jainam/catkin_ws/src/test3/meshes/wheel_x5.dae"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="${pi} 0 0"/>
        <geometry>
          <mesh filename="/home/jainam/catkin_ws/src/test3/meshes/wheel_x5.dae"/>
        </geometry>
      </collision>
      <xacro:cylinder_inertia m="5" r="${wheel_radius}" h="0.085"/>
    </link>

    <joint name="right_wheel_joint1" type="continuous">
      <axis xyz="0 1 0" rpy="0 0 0" />
      <parent link="chassis"/>
      <child link="right_wheel"/>
      <origin xyz="${-wheel_joint_offset_x} -${wheel_joint_offset_y} -0.038441" rpy="0 0 ${pi/2}"/>
    </joint>

  <link name="caster_wheel">
    <visual>
      <origin xyz="0 0 0 " rpy="0 0 0"/>
      <geometry>
        <sphere radius="${caster_wheel_radius}"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <sphere radius="${caster_wheel_radius}"/>      
      </geometry>
    </collision>
    <xacro:sphere_inertia m="2" r="${caster_wheel_radius}"/>
  </link>
  
<!--Caster Wheel Joint-->
  <joint name="caster_wheel_joint" type="continuous">
    <axis xyz="0 1 0" rpy="0 0 0" />
    <parent link="chassis"/>
    <child link="caster_wheel"/>
    <origin xyz="0 ${caster_wheel_joint_offset + 0.009} ${-base_height/2 - 0.0225}" rpy="0 0 0"/>
  </joint>

  <link name="aruco_marker">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="/home/jainam/catkin_ws/src/test3/meshes/aruco_marker_x5.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="/home/jainam/catkin_ws/src/test3/meshes/aruco_marker_x5.dae"/>  
      </geometry>
    </collision>
    <xacro:box_inertia m="0.001" w="0.2209" h="0.2209" d="0.0025"/>
  </link>

  <joint name="aruco_marker_joint" type="fixed">
    <parent link="chassis"/>
    <child link="aruco_marker"/>
    <origin xyz="0 0 ${base_height/2 }" rpy="0 0 0"/>
  </joint>

  <gazebo>
    <plugin name="differential_drive_controller" filename="libgazebo_ros_diff_drive.so">
      <alwaysOn>false</alwaysOn>
      <legacyMode>false</legacyMode>
      <updateRate>20</updateRate>
      <leftJoint>left_wheel_joint1</leftJoint>
      <rightJoint>right_wheel_joint1</rightJoint>
      <wheelSeparation>${wheel_separation}</wheelSeparation>
      <wheelDiameter>${wheel_radius * 2}</wheelDiameter>
      <torque>20</torque>
      <commandTopic>/tbot/base_controller/cmd_vel</commandTopic>
      <odometryTopic>/tbot/base_controller/odom</odometryTopic>
      <odometryFrame>odom</odometryFrame>
      <robotBaseFrame>base_footprint1</robotBaseFrame>
    </plugin>
  </gazebo>

</robot>